### PR TITLE
fix race on _storage field.

### PR DESCRIPTION
**Pull request by heejaechang**
**Customer scenario**

OOP randomly crash and bring down VS.

**Bugs this fixes:**

https://github.com/dotnet/roslyn/issues/18697

**Workarounds, if any**

There is no workaround

**Risk**

Low. 

no behavior change. code is changed to use explicit lock to remove race.

**Performance impact**

Low. individual instance has its own lock and access to it shouldn't have too many traffic.

**Is this a regression from a previous update?**

No

**Root cause analysis:**

This race seems been there for a long time. but we were lucky. it happened for a while and suddenly no longer repro. but we want to harden this code path.

**How was the bug found?**

Dogfooding.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs [hunk 1]
```diff
@@ -92,7 +92,8 @@ public RecoverableSyntaxRoot<TRoot> WithSyntaxTree(IRecoverableSyntaxTree<TRoot>
                 if (this.TryGetValue(out var root))
                 {
                     var result = new RecoverableSyntaxRoot<TRoot>(_service, root, containingTree);
-                    result._storage = _storage;
+                    Interlocked.CompareExchange(ref result._storage, _storage, null);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: can we avoid complex interlockeds and just use normal locks?

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs [hunk 2]
```diff
@@ -101,18 +102,20 @@ public RecoverableSyntaxRoot<TRoot> WithSyntaxTree(IRecoverableSyntaxTree<TRoot>
...
-                    root.SerializeTo(stream, cancellationToken);
+                    root.SerializeTo(stream, CancellationToken.None);
                     stream.Position = 0;
 
-                    _storage = _service.LanguageServices.WorkspaceServices.GetService<ITemporaryStorageService>().CreateTemporaryStreamStorage(cancellationToken);
-                    await _storage.WriteStreamAsync(stream, cancellationToken).ConfigureAwait(false);
+                    var storage = _service.LanguageServices.WorkspaceServices.GetService<ITemporaryStorageService>().CreateTemporaryStreamStorage(CancellationToken.None);
+                    await storage.WriteStreamAsync(stream, CancellationToken.None).ConfigureAwait(false);
+
+                    Interlocked.CompareExchange(ref _storage, storage, null);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: same comment.  why not just have a SemaphoreSlim here.  We're already async.  If two callers come in, i don't see why we want both to write teh stream, and then one to succeed replacing the storage pointer.

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs (hunk 1 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: if we go for simplicity, I can change the variable to volatile.

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs (hunk 2 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: the caller of this method make sure there is ever only 1 calls for Save.

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: Please never use volatile.

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: Regardless, please do not use complex concurrency constructs when simple ones will do.  It makes the code much harder to review and much harder to maintain.

**Comment by heejaechang** [comment]
@Pilchie @jasonmalinowski have you tried the fix?

**Comment by Pilchie** [comment]
I don't have a machine where I'm seeing the problem right now :(

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs [hunk 3]
```diff
@@ -62,7 +62,7 @@ internal SyntaxTreeInfo WithOptionsAndLength(ParseOptions options, int length)
         internal sealed class RecoverableSyntaxRoot<TRoot> : RecoverableWeakValueSource<TRoot>
             where TRoot : SyntaxNode
         {
-            private ITemporaryStreamStorage _storage;
+            private volatile ITemporaryStreamStorage _storage;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: Please don't use volatile.  This is becoming complex in a manner that has not been given justificaiton.

**Comment by CyrusNajmabadi** [comment]
Please either change to this be a simple lock.  Or, if there is a belief there will be high contention, and we need asynchrony, a SemaphoreSlim.

**Comment by jasonmalinowski** [comment]
@heejaechang I haven't seen any failures without this fix, so I'm not sure I can make a meaningful test. I'm OK taking the fix either way -- it does look legitimately bad and it's our best theory for what we saw. I do agree with @CyrusNajmabadi though -- let's just use a simple lock.

**Comment by sharwell** [comment]
@heejaechang Can you describe the problem this is intending to fix? Right now I'm missing context.

**Comment by heejaechang** [comment]
okay. I made code to use explicit lock.

**Comment by heejaechang** [comment]
@sharwell some people saw OOP crashing due to SyntaxTree recovery threw an exception. it was happening since ITemproaryStreamStorage was null which can't happen if SyntaxTree is ever saved. and this behavior seems random meaning looks like a race and this was only thing I can spot that can cause this kind of race.

**Comment by CyrusNajmabadi** [comment]
So, i don't know why this would necessarily help.  But i don't see this hurting.  What is the reason we think we need this?

**Comment by heejaechang** [comment]
@CyrusNajmabadi jason and kevin saw random OOP crash (which cause VS to crash) due to SyntaxTree recovery throwing an exception. it was throwing exception because temporary storage that must exist didn't exist. this fix is to fix potential race which could cause this issue.

**Comment by sharwell** [comment]
@heejaechang Do you have a stack trace from this situation?

**Comment by heejaechang** [comment]
@sharwell this is throwing
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs,131

this is one of callstack thrown by Recover of recoverable tree

Roslyn.Utilities.Contract.ThrowIfNull[T](T,String) Microsoft.CodeAnalysis.Host.AbstractSyntaxTreeFactoryService.RecoverableSyntaxRoot`1.Recover(CancellationToken) Roslyn.Utilities.AsyncLazy`1.GetValue(CancellationToken) Microsoft.CodeAnalysis.Host.RecoverableWeakValueSource`1.GetValue(CancellationToken) Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTreeFactoryServiceFactory.CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.GetRoot(CancellationToken) Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTree.GetDiagnostics(CancellationToken) Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnostics(CompilationStage,Boolean,CancellationToken) Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods(CommonPEModuleBuilder,Boolean,DiagnosticBag,Predicate`1,CancellationToken) Microsoft.CodeAnalysis.Compilation.Emit(Stream,Stream,Stream,Stream,IEnumerable`1,EmitOptions,IMethodSymbol,Stream,IEnumerable`1,CompilationTestData,CancellationToken) Microsoft.CodeAnalysis.Compilation.Emit(Stream,Stream,Stream,Stre...

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure is this known issue?
Roslyn.VisualStudio.IntegrationTests.VisualBasic.BasicCodeActions.GenerateMethodInClosedFile [FAIL]

https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/5282/

**Comment by heejaechang** [comment]
retest windows_debug_vs-integration_prtest  please

**Comment by heejaechang** [comment]
ping? can I get some sign off? @dotnet/roslyn-analysis @dotnet/roslyn-ide 

also, should I take this to 15.3? or put it in 15.6 would be fine?

**Comment by sharwell** [comment]
@heejaechang I'm still looking at this. I'll finish shortly.

**Comment by sharwell** [comment]
@heejaechang I am not convinced this code would produce a race condition with respect to the `_storage` field as described (i.e. in commit dc1fd90). However, there appears to be a *much* more likely race condition which this pull request does not correct.

1. Thread 1 calls `GetValue` and proceeds to the entry point of `ResetRecoverySource`
2. Thread 2 calls `RecoverableSyntaxRoot<TRoot>.WithSyntaxTree`
    * `this.TryGetValue` returns true because the instance was set in
    * `this._storage` is still null

When the GC later reclaims the weakly-referenced instance, the object produced by Thread 2 will be in a bad state - the value source will expect `_storage` to be non-null but it will still be null.

:memo: I don't see any reason to add additional locks here. I would prefer to see this PR start over from master and correct the true race condition if possible without placing additional locks on the fast path.

**Comment by heejaechang** [comment]
@sharwell this remove the race you are concerning about.

http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Utilities/ValuesSources/RecoverableWeakValueSource.cs,123

**Comment by heejaechang** [comment]
@sharwell the race we are concerning about is read write ordering between multiple threads. without memory barrier, even if one thread write a field, there is no guarantee other thread will see the write, the lock is there so that we have memory barrier so that all read see all writes from other threads.

**Comment by sharwell** [comment]
> this remove the race you are concerning about.

That line ensures the call to `WithSyntaxTree` will go through the first code path, resulting in the invalid state if the save operation is not yet complete.

> without memory barrier, even if one thread write a field, there is no guarantee other thread will see the write, the lock is there so that we have memory barrier so that all read see all writes from other threads.

Assuming that the JIT doesn't reorder the writes to `_storage` and `_recoverySource` (which it shouldn't, since they are in two different TPL tasks), the only way this could occur in the processor is if they are split across cache lines. Since the fields appear consecutively in memory, this is exceedingly unlikely.

**Comment by heejaechang** [comment]
@sharwell ya. the result._storage seems completely wrong since it doesn't reset _saved field. if it was ever called after saved is done, VS would have crashed due to this - http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs,106

but I dont see any crash due to it. so no idea how it worked before.

> That line ensures the call to WithSyntaxTree will go through the first code path, resulting in the invalid state if the save operation is not yet complete.

I think you got it reserved. it is not a problem if save is not done yet since _storage will be null which is expected. it is problem is saved is happened. since code expect _storage to be null.

...

if you can come up with any other theory how _storage could be null, then I am welcome to consider. if WithSyntaxTree went to the first code path, Recover will be never get called since root is alive in memory.

**Comment by sharwell** [comment]
@heejaechang `WithSyntaxTree` is producing an object with the following characteristics:

1. `_valueSource` is an `AsyncLazy<T>` which calls `Recover` to obtain the value
1. `_saved` is true
1. `_weakInstance` is a weak reference to the target
1. `_storage` is null

As long as `_weakInstance` keeps its reference alive, this is fine. However, down the road when the GC reclaims that object (some time after the last `GC.KeepAlive`), this object will be in a state where `Recover` will be called with `_storage==null` the next time its value is requested.

**Comment by heejaechang** [comment]
if call get into this
 if (this.TryGetValue(out var root))
                {
                    var result = new RecoverableSyntaxRoot<TRoot>(_service, root, containingTree);
                    result._storage = _storage;
                    return result;
                }

it create new recoverable root of this

 public RecoverableSyntaxRoot(
                AbstractSyntaxTreeFactoryService service,
                TRoot root,
                IRecoverableSyntaxTree<TRoot> containingTree)
                : base(new ConstantValueSource<TRoot>(containingTree.CloneNodeAsRoot(root)))
            {
                _service = service;
                _containingTree = containingTree;
            }

which give the base constant value source
public RecoverableWeakValueSource(ValueSource<T> initialValue)
        {
            _weakInstance = s_noReference;
            _recoverySource = initialValue;
        }

which will have _saved set to false, which will try to save the root. which is this code

 protected override async Task SaveAsync(TRoot root, CancellationToken cancellationToken)
            {
                Contract.ThrowIfFalse(_storage == null); // Cannot save more than once
 
                // tree will be always held alive in memory, but nodes come and go. serialize nodes to storage
                using (var stream = SerializableBytes.CreateWritableStream())
                {
                    root.SerializeTo(stream, cancellationToken);
                    stream.Position = 0;
 
                    _storage = _service.LanguageServices.WorkspaceServices.GetService<ITemporaryStorageService>().CreateTemporaryStreamStorage(cancellationToken);
                    await _storage.WriteStreamAsync(stream, cancellationToken).ConfigureAwait(false);
                }
            }

which expect storage to be null.

until this point, the root should be alive. 

I am not sure the state you mentioned above, how it can happen? I think you are referring second code path in WithSyntaxTree but with that code path, _storage must be always not null.

**Comment by sharwell** [comment]
@heejaechang For some reason I was thinking the first code path went through the other constructor. The second code path is the one that sets `_saved` to true, which means the conditions for this to occur are different:

* The current instance has *not* been saved
* The argument `containingTree` has been saved, but not yet flushed (the path described by my "Thread 1" earlier)

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure ETA test seems busted? is this known issue? -https://ci.dot.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_eta_open_prtest/7290/

**Comment by heejaechang** [comment]
@sharwell so how can _storage be null when root is not alive in memory? I linked in my first comment that due to KeepAlive, second path can't be got into unless SaveAsync has happened which will set _storage.

**Comment by jasonmalinowski** [comment]
Machine was in a broken state and the screenshot test ETA does to validate screenshot can be taken failed.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest windows_eta_open_prtest please.

**Comment by heejaechang** [comment]
@sharwell what change are you asking me to make?

**Comment by sharwell** [comment]
> so how can _storage be null when root is not alive in memory?

When `GetValue` has never been called on the current object. Even though the first constructor uses a `ConstantValueSource` so the root is alive in memory, it initializes `_weakInstance`  to `s_noReference` so `TryGetValue` will return false.

**Comment by sharwell** [comment]
> what change are you asking me to make?

It's the end of a long day and I already misread this once. Let me look tomorrow morning with fresh eyes and get back to you.

**Comment by heejaechang** [comment]
@sharwell actually I think you spotted the bug. TryGetValue just checking weakReference is wrong. I should check but weakReference and recoverable source. not sure how this bug survived this long...

**Comment by heejaechang** [comment]
can I get code review? @dotnet/roslyn-analysis @dotnet/roslyn-ide @sharwell spotted the bug where we didn't check all sources for root and went into wrong code path.

File: src/Workspaces/Core/Portable/Utilities/ValuesSources/RecoverableWeakValueSource.cs [hunk 4]
```diff
@@ -65,7 +65,8 @@ public RecoverableWeakValueSource(RecoverableWeakValueSource<T> savedSource)
 
         public override bool TryGetValue(out T value)
         {
-            return _weakInstance.TryGetTarget(out value);
+            return _weakInstance.TryGetTarget(out value) ||
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Utilities/ValuesSources/RecoverableWeakValueSource.cs]: can this get any documentation to explain why it is correct?

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs [hunk 5]
```diff
@@ -91,9 +91,7 @@ public RecoverableSyntaxRoot<TRoot> WithSyntaxTree(IRecoverableSyntaxTree<TRoot>
             {
                 if (this.TryGetValue(out var root))
                 {
-                    var result = new RecoverableSyntaxRoot<TRoot>(_service, root, containingTree);
-                    result._storage = _storage;
-                    return result;
+                    return new RecoverableSyntaxRoot<TRoot>(_service, root, containingTree);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: why was the previous code wrong?

**Comment by CyrusNajmabadi** [comment]
Sorry, i need more data.  Just trying to read the code for RecoverableWeakValueSource and i have no real idea about how it is supposed to work or why the new code is correct.  I would not feel comfortable trying to come in and fix any others we might see here in the future.

File: src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs (hunk 5 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs]: this explain why it is wrong. https://github.com/dotnet/roslyn/pull/20051#issuecomment-307260122

File: src/Workspaces/Core/Portable/Utilities/ValuesSources/RecoverableWeakValueSource.cs (hunk 4 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Utilities/ValuesSources/RecoverableWeakValueSource.cs]: sure.

**Comment by sharwell** [comment]
@heejaechang Since the readability of this code is unlikely to substantially improve, can you add a regression test which fails prior to this change but passes afterwards?

**Comment by heejaechang** [comment]
added test

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs [hunk 6]
```diff
@@ -1465,6 +1465,33 @@ public void TestProjectCompletenessWithMultipleProjects()
             }
         }
 
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void TestWithSyntaxTree()
```
Comment (sharwell) [inline src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs]: :bulb:

```csharp
[WorkItem(18697, "https://github.com/dotnet/roslyn/issues/18697")]
```

**Comment by heejaechang** [comment]
@CyrusNajmabadi I added more comment. also added unit test. any other one who want to take a look?

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs [hunk 7]
```diff
@@ -1465,6 +1465,34 @@ public void TestProjectCompletenessWithMultipleProjects()
...
+                strongTree.Options, 
+                new ConstantValueSource<TextAndVersion>(TextAndVersion.Create(strongTree.GetText(), VersionStamp.Create(), strongTree.FilePath)), 
+                strongTree.GetText().Encoding, 
+                strongTree.GetRoot());
+
+            // create new tree before it ever getting root node
+            var newTree = recoverableTree.WithFilePath("different/dummy");
+
+            // this shouldn't throw
+            var root = newTree.GetRoot();
```
Comment (heejaechang) [inline src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs]: without fix, this throws.

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest please

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure is there a known issue on windows_release_vs-integration_prtest?

**Comment by ivanbasov** [comment]
yes, it is a known issue. @jasonmalinowski is going to temporarily remove windows_release_vs-integration_prtest from the list of required ones.

**Comment by jasonmalinowski** [comment]
@heejaechang (and @ivanbasov as FYI): integration tests are temporarily marked as not required so you'll be able to merge.

**Comment by jasonmalinowski** [comment]
However, this still needs shiproom approval. Do we have that?

**Comment by heejaechang** [comment]
@jasonmalinowski I dont know whether we want it for 15.3 or not. @Pilchie @jinujoseph ?

**Comment by Pilchie** [comment]
I haven't hit it again since that one day when it completely blocked me.  

@tmeschter Any chance we can look through Watson telemetry to get a sense of how often this hits?  @heejaechang The issue exists in the public 15.1 and 15.2 releases, right?

**Comment by tmeschter** [comment]
@Pilchie The call stack in question is the one listed in #18697?

**Comment by Pilchie** [comment]
@tmeschter Added you to an internal thread.  I think there were a couple of stacks involved.

**Comment by heejaechang** [comment]
@Pilchie the issue should be in since dev14. the race is general Workspace issue.

**Comment by heejaechang** [comment]
@tmeschter callstack is this.

Roslyn.Utilities.Contract.ThrowIfNull[T](T,String) Microsoft.CodeAnalysis.Host.AbstractSyntaxTreeFactoryService.RecoverableSyntaxRoot`1.Recover(CancellationToken) Roslyn.Utilities.AsyncLazy`1.GetValue(CancellationToken) Microsoft.CodeAnalysis.Host.RecoverableWeakValueSource`1.GetValue(CancellationToken) Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTreeFactoryServiceFactory.CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.GetRoot(CancellationToken) Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTree.GetDiagnostics(CancellationToken) Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnostics(CompilationStage,Boolean,CancellationToken) Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods(CommonPEModuleBuilder,Boolean,DiagnosticBag,Predicate`1,CancellationToken) Microsoft.CodeAnalysis.Compilation.Emit(Stream,Stream,Stream,Stream,IEnumerable`1,EmitOptions,IMethodSymbol,Stream,IEnumerable`1,CompilationTestData,CancellationToken) Microsoft.CodeAnalysis.Compilation.Emit(Stream,Stream,Stream,Stre...

**Comment by tmeschter** [comment]
Hmm, I can't find that issue in Watson. It may be that it's getting over-bucketed with a bunch of other crashes in the remote process.

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest  please

**Comment by heejaechang** [comment]
Retest windows_release_vs-integration_prtest please

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure test failed 
Roslyn.VisualStudio.IntegrationTests.VisualBasic.BasicCodeActions.GenerateMethodInClosedFile

https://ci.dot.net/job/dotnet_roslyn/job/dev15.6/job/windows_release_vs-integration_prtest/495/

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest please

**Comment by cston** [comment]
`BasicCodeActions.GenerateMethodInClosedFile` failure is #20371.

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure where should I check this in? dev15.6 or master?

**Comment by sharwell** [comment]
@heejaechang master branch is currently for 15.5 release.

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure integration test failed something related to vsix extension installation
https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/5664/

**Comment by heejaechang** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest please

**Comment by heejaechang** [comment]
close and re open to run tests again.

**Comment by heejaechang** [comment]
retest windows_debug_vs-integration_prtest  please

**Pull request merged**